### PR TITLE
Add Basic sanity cloud migration integration tests

### PR DESCRIPTION
**Pull request by ram-infrac**
This PR add base supports for cloud migration integration tests, includes following changes -
     - read configmap for cluster2 kubeconfig file and dump it to /tmp/kubeconfig location
     - read cluster pair info & create clusterpair file
     - apply clusterpair file and validate same
    - schedule mysql-1-pvc  pods and get context(ctx)
    - apply & migration spec from (integration_test/specs)
    - wait for application to come in running state on cluster2 

**Note** : this also include changes in from torpedo pr https://github.com/portworx/torpedo/pull/217

### CODE REVIEW

File: test/integration_test/spec_template.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+kind: ClusterPair
+metadata:
+  name: {{ .PairName }}
+spec:
+  options:
+    ip: {{ .RemoteIP }}
+    token: {{ .RemoteToken }} 
+    port: "{{ .RemotePort }}"
+  config:
+    clusters:
```
Comment (ram-infrac) [inline test/integration_test/spec_template.go]: @disrani-px  I think this section(user-context) is not constant we should also include this as part of template variables i.e{user-context}
storkctl generate also assumes user-context section is constant. Please let me know which on to proceed with

File: test/integration_test/common_test.go [hunk 2]
```diff
@@ -4,20 +4,30 @@ package integrationtest
...
+	"io/ioutil"
+	"os"
+	"strconv"
 	"testing"
 	"time"
 
 	storkdriver "github.com/libopenstorage/stork/drivers/volume"
 	_ "github.com/libopenstorage/stork/drivers/volume/portworx"
+	"github.com/portworx/sched-ops/k8s"
+	k8s_ops "github.com/portworx/sched-ops/k8s"
```
Comment (disrani-px) [inline test/integration_test/common_test.go]: Imported twice

File: test/integration_test/spec_template.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+	RemoteConfigCertData string
+}
+
+var clusterPairSpec = `apiVersion: stork.libopenstorage.org/v1alpha1
+kind: ClusterPair
+metadata:
+  name: {{ .PairName }}
+spec:
+  options:
+    ip: {{ .RemoteIP }}
```
Comment (disrani-px) [inline test/integration_test/spec_template.go]: These are PX specific, they shouldn't be here. You should get the options through torpedo APIs and just add the options map here

File: test/integration_test/migrs/migration.yaml [hunk 4]
```diff
@@ -0,0 +1,14 @@
+apiVersion: stork.libopenstorage.org/v1alpha1
```
Comment (disrani-px) [inline test/integration_test/migrs/migration.yaml]: The specs should go under the specs directory. Also use a more appropriate name for the migration

File: test/integration_test/cloud_migration_test.go [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+import (
+	"testing"
+	"time"
+
+	"github.com/portworx/torpedo/drivers/scheduler"
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/require"
+)
+
+func testBasicCloudMigration(t *testing.T) {
```
Comment (disrani-px) [inline test/integration_test/cloud_migration_test.go]: Don't use CloudMigration. The spec and controllers are called Migration. Rename the test file too.

File: test/integration_test/cloud_migration_test.go [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+	err = CreateClusterPairSpec(specReq)
+	require.NoError(t, err, "Error creating cluster Spec")
+
+	ctx, err := getContextCRD("cluster-pair")
+	require.NoError(t, err, "Error locating cluster Spec")
+
+	err = schedulerDriver.CreateCRDObjects(ctx, 2*time.Minute, 10*time.Second)
+	require.NoError(t, err, "Error applying clusterpair")
+
```
Comment (disrani-px) [inline test/integration_test/cloud_migration_test.go]: Missing WaitForRunning()

File: test/integration_test/common_test.go [hunk 7]
```diff
@@ -81,6 +94,7 @@ func TestMain(t *testing.T) {
 	t.Run("HealthMonitor", testHealthMonitor)
 	t.Run("Snapshot", testSnapshot)
 	t.Run("CmdExecutor", asyncPodCommandTest)
+	t.Run("testBasicMigartion", testBasicMigration)
```
Comment (disrani-px) [inline test/integration_test/common_test.go]: t.Run("Migartion", testMigration)

File: test/integration_test/common_test.go [hunk 8]
```diff
@@ -201,6 +215,102 @@ func verifyScheduledNode(t *testing.T, appNode node.Node, volumes []string) {
...
 }
 
+// write kubbeconfig file to  /tmp/kubeconfig
+func dumpRemoteKubeConfig(configObject string) error {
+	cm, err := k8s_ops.Instance().GetConfigMap(configObject, "kube-system")
+	if err != nil {
+		logrus.Errorf("Error reading config map: %v", err)
+		return err
+	}
+	status := cm.Data["kubeconfig"]
```
Comment (disrani-px) [inline test/integration_test/common_test.go]: This isn't the status, right?

File: test/integration_test/common_test.go [hunk 9]
```diff
@@ -201,6 +215,102 @@ func verifyScheduledNode(t *testing.T, appNode node.Node, volumes []string) {
...
+	if len(status) == 0 {
+		logrus.Info("found empty failure status for key:remoteConifg in config map")
```
Comment (disrani-px) [inline test/integration_test/common_test.go]: Log as error. Also this isn't the status, so log message needs to be re-worded.

File: test/integration_test/migration_test.go [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+	require.NoError(t, err, "Error creating cluster Spec")
+
+	err = schedulerDriver.RescanSpecs(specDir)
+	require.NoError(t, err, "Unable to parse spec dir")
+
+	_, err = schedulerDriver.Schedule("clusterpair",
+		scheduler.ScheduleOptions{AppKeys: []string{"cluster-pair"}})
+	require.NoError(t, err, "Error applying clusterpair")
+
+	logrus.Info("Validated Cluster Pair Specs")
```
Comment (disrani-px) [inline test/integration_test/migration_test.go]: Put the steps till here in a separate function so different test can call it to set up the cluster pair.

File: test/integration_test/migration_test.go [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+	//  schedule mysql app
+	ctxs, err := schedulerDriver.Schedule("singlemysql",
+		scheduler.ScheduleOptions{AppKeys: []string{"mysql-1-pvc"}})
+	require.NoError(t, err, "Error scheduling task")
+	require.Equal(t, 1, len(ctxs), "Only one task should have started")
+
+	err = schedulerDriver.WaitForRunning(ctxs[0], defaultWaitTimeout, defaultWaitInterval)
+	require.NoError(t, err, "Error waiting for pod to get to running state")
+
+	_, err = schedulerDriver.Schedule("migrs",
```
Comment (disrani-px) [inline test/integration_test/migration_test.go]: migration instead of migrs

File: test/integration_test/migration_test.go [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+	require.NoError(t, err, "Error setting remote config")
+
+	err = schedulerDriver.WaitForRunning(ctxs[0], defaultWaitTimeout, defaultWaitInterval)
+	require.NoError(t, err, "Error waiting for pod to get to running state")
+	destroyAndWait(t, ctxs)
+
+	// destroy mysql app on cluster 1
+	err = setRemoteConfig("")
+	require.NoError(t, err, "Error setting remote config")
+	destroyAndWait(t, ctxs)
```
Comment (disrani-px) [inline test/integration_test/migration_test.go]: Need to destroy on the destination cluster too.

File: test/integration_test/specs/migration/migration.yaml [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+spec:
+  # This should be the name of the cluster pair created above
+  clusterPair: remoteclusterpair
+  # If set to false this will migrate only the Portworx volumes. No PVCs, apps, etc will be migrated
+  includeResources: true
+  # If set to false, the deployments and stateful set replicas will be set to 0 on the destination. There will be an annotation with "stork.openstorage.org/migrationReplicas" to store the replica count from the source
+  startApplications: true
+  # List of namespaces to migrate
+  namespaces:
+  - mysql-1-pvc-singlemysql
```
Comment (disrani-px) [inline test/integration_test/specs/migration/migration.yaml]: How does this work? torpedo generates a random name for the namespace

Comment (ram-infrac) [reply test/integration_test/specs/migration/migration.yaml]: torpedo does not generates random name, we supply it while scheduling using **generateInstanceID** . I'm simply using **singlemysql** as instanceID ,which torpedo then forms namespace as  **mysql-1-pvc-singlemysql**  (i.e specKey-instanceID)

File: test/integration_test/migration_test.go (hunk 12 above)
Comment (ram-infrac) [reply test/integration_test/migration_test.go]: we do while after waitforrunning in cluster 2,added comment to clarify

File: test/integration_test/common_test.go (hunk 8 above)
Comment (ram-infrac) [reply test/integration_test/common_test.go]: right its config value string

File: test/integration_test/migration_test.go [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+func sanityMigrationTest(t *testing.T) {
+	var err error
+	// create, apply and validate cluster pair specs
+	pairCtxs, err := scheduleClusterPair()
+	require.NoError(t, err, "Error scheduling cluster pair")
+
+	// schedule mysql app on cluster 1
+	mySQLCtxs, err := schedulerDriver.Schedule("singlemysql",
```
Comment (disrani-px) [inline test/integration_test/migration_test.go]: mysqlCtxs

File: test/integration_test/migration_test.go [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+		scheduler.ScheduleOptions{AppKeys: []string{"mysql-1-pvc"}})
+	require.NoError(t, err, "Error scheduling task")
+	require.Equal(t, 1, len(mySQLCtxs), "Only one task should have started")
+
+	err = schedulerDriver.WaitForRunning(mySQLCtxs[0], defaultWaitTimeout, defaultWaitInterval)
+	require.NoError(t, err, "Error waiting for pod to get to running state")
+
+	// apply migration specs
+	migrsCtxs, err := schedulerDriver.Schedule("migration",
```
Comment (disrani-px) [inline test/integration_test/migration_test.go]: migrationCtxs

File: test/integration_test/migration_test.go [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+import (
+	"testing"
+
+	"github.com/portworx/torpedo/drivers/scheduler"
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/require"
+)
+
+func testMigration(t *testing.T) {
+	logrus.Info("Basic sanity tests for cloud migration")
```
Comment (disrani-px) [inline test/integration_test/migration_test.go]: Remove the log

File: test/integration_test/common_test.go [hunk 17]
```diff
@@ -201,6 +215,151 @@ func verifyScheduledNode(t *testing.T, appNode node.Node, volumes []string) {
...
+func dumpRemoteKubeConfig(configObject string) error {
+	cm, err := k8s_ops.Instance().GetConfigMap(configObject, "kube-system")
+	if err != nil {
+		logrus.Errorf("Error reading config map: %v", err)
+		return err
+	}
+	config := cm.Data["kubeconfig"]
+	if len(config) == 0 {
+		configErr := "found empty remoteConfig in config map"
+		logrus.Error("Error reading kubeconfig %s", configErr)
```
Comment (disrani-px) [inline test/integration_test/common_test.go]: You can put the complete error message here and use it below

**Pull request merged**